File: lecture_03/matlab/block2_noise_statistics.m
%% BLOCK 2 — Noise, Signal Statistics & Basic Signal Processing
% Lecture 3: MATLAB Summer Prep | NSBV BC2001
% -----------------------------------------------
% Real neuroscience data is always noisy.
% EEG and electrophysiology recordings contain both signal and noise.
% Here we simulate that and learn to characterise what we have.

clc; clearvars;

%% 1. Adding Gaussian noise to a clean signal
t      = linspace(0, 1, 1000);
clean  = sin(2 * pi * 10 * t);         % clean 10 Hz signal

noise_level = 0.5;
noise       = noise_level * randn(size(clean));   % randn = random normal (Gaussian)
noisy       = clean + noise;

figure;
subplot(2,1,1);
plot(t, clean, 'b-', 'LineWidth', 1.5);
xlabel('Time (s)'); ylabel('Amplitude');
title('Clean Signal');

subplot(2,1,2);
plot(t, noisy, 'Color', [0.6 0.6 0.6], 'LineWidth', 0.8);
hold on;
plot(t, clean, 'b-', 'LineWidth', 1.5);
hold off;
xlabel('Time (s)'); ylabel('Amplitude');
title('Noisy Signal (grey) vs Clean (blue)');
legend('Noisy', 'Clean');

%% 2. Descriptive statistics of the signal
fprintf('--- Signal Statistics ---\n');
fprintf('Mean:     %.4f\n', mean(noisy));
fprintf('Std Dev:  %.4f\n', std(noisy));
fprintf('Min:      %.4f\n', min(noisy));
fprintf('Max:      %.4f\n', max(noisy));
fprintf('Range:    %.4f\n', max(noisy) - min(noisy));

%% 3. Simple moving average (smoothing)
% Smoothing reduces noise — fundamental technique in EEG preprocessing
window_size = 20;                          % samples
smoothed    = movmean(noisy, window_size); % built-in moving average

figure;
plot(t, noisy,    'Color', [0.8 0.8 0.8], 'LineWidth', 0.8);
hold on;
plot(t, smoothed, 'r-', 'LineWidth', 2);
plot(t, clean,    'b--', 'LineWidth', 1.5);
hold off;
xlabel('Time (s)'); ylabel('Amplitude');
title('Noisy vs Smoothed vs Clean Signal');
legend('Noisy', 'Smoothed', 'Clean');

%% 4. Signal-to-Noise Ratio (SNR)
signal_power = mean(clean .^ 2);
noise_power  = mean(noise .^ 2);
snr_val      = 10 * log10(signal_power / noise_power);
fprintf('\nSignal-to-Noise Ratio: %.2f dB\n', snr_val);

%% 5. Histogram of signal values
figure;
histogram(noisy, 40, 'FaceColor', [0.4 0.6 0.9], 'EdgeColor', 'white');
xlabel('Amplitude'); ylabel('Count');
title('Distribution of Signal Values');
xline(mean(noisy), 'r--', 'LineWidth', 2, 'Label', 'Mean');

File: lecture_03/python/block2_noise_statistics.py
# BLOCK 2 — Noise, Signal Statistics & Basic Signal Processing (Python)
# Lecture 3: MATLAB Summer Prep | NSBV BC2001
# -----------------------------------------------
# Python parallel: compare with block2_noise_statistics.m

import numpy as np
import matplotlib.pyplot as plt

t     = np.linspace(0, 1, 1000)
clean = np.sin(2 * np.pi * 10 * t)

# --- 1. Adding Gaussian noise ---
# MATLAB: randn(size(clean))  →  Python: np.random.randn(*clean.shape)
noise_level = 0.5
noise       = noise_level * np.random.randn(*clean.shape)
noisy       = clean + noise

fig, axes = plt.subplots(2, 1, figsize=(10, 6))
axes[0].plot(t, clean, 'b-', linewidth=1.5)
axes[0].set_title('Clean Signal'); axes[0].set_ylabel('Amplitude')

axes[1].plot(t, noisy, color='0.6', linewidth=0.8, label='Noisy')
axes[1].plot(t, clean, 'b-',        linewidth=1.5, label='Clean')
axes[1].set_title('Noisy vs Clean'); axes[1].set_ylabel('Amplitude')
axes[1].set_xlabel('Time (s)'); axes[1].legend()
plt.tight_layout(); plt.show()

# --- 2. Descriptive statistics ---
print('--- Signal Statistics ---')
print(f'Mean:    {np.mean(noisy):.4f}')
print(f'Std Dev: {np.std(noisy):.4f}')
print(f'Min:     {np.min(noisy):.4f}')
print(f'Max:     {np.max(noisy):.4f}')
print(f'Range:   {np.ptp(noisy):.4f}')   # ptp = peak-to-peak (max - min)

# --- 3. Simple moving average (smoothing) ---
# NumPy doesn't have movmean built-in; use np.convolve for a clean implementation
def moving_average(x, w):
    """Compute a centred moving average of window size w."""
    return np.convolve(x, np.ones(w) / w, mode='same')

window_size = 20
smoothed    = moving_average(noisy, window_size)

plt.figure(figsize=(10, 4))
plt.plot(t, noisy,    color='0.8', linewidth=0.8, label='Noisy')
plt.plot(t, smoothed, 'r-',        linewidth=2,   label='Smoothed')
plt.plot(t, clean,    'b--',       linewidth=1.5, label='Clean')
plt.xlabel('Time (s)'); plt.ylabel('Amplitude')
plt.title('Noisy vs Smoothed vs Clean'); plt.legend()
plt.show()

# --- 4. Signal-to-Noise Ratio ---
signal_power = np.mean(clean ** 2)
noise_power  = np.mean(noise ** 2)
snr_val      = 10 * np.log10(signal_power / noise_power)
print(f'\nSignal-to-Noise Ratio: {snr_val:.2f} dB')

# --- 5. Histogram ---
plt.figure(figsize=(7, 4))
plt.hist(noisy, bins=40, color=[0.4, 0.6, 0.9], edgecolor='white')
plt.xlabel('Amplitude'); plt.ylabel('Count')
plt.title('Distribution of Signal Values')
plt.axvline(np.mean(noisy), color='red', linestyle='--', linewidth=2, label='Mean')
plt.legend(); plt.show()
